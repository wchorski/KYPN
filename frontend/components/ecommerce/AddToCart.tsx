'use client'
import { useMutation, gql } from '@apollo/client';
import { useRouter } from 'next/navigation';
import { useSession } from "../menus/Session";
import { MdShoppingBag } from 'react-icons/md';
import { TbCheck, TbExclamationCircle, TbQuestionMark, TbLoader  } from 'react-icons/tb';
import { QUERY_USER_CURRENT } from '../menus/Session';
import { useState } from 'react';
import styles from '../../styles/eyecandy/SpinCycle.module.scss'
import stylesButton from "@/styles/ecommerce/AddToCart.module.scss";

const delay = (ms:number) => new Promise(res => setTimeout(res, ms));

type State = 'loading'|'pending'|'error'|'out_of_stock'|'success'|undefined

export default function AddToCart({ id }: { id: string }) {

  const [state, setstate] = useState<State>(undefined)
  const session = useSession()
  const router = useRouter()
  const [addToCart, { loading }] = useMutation(ADD_TO_CART_MUTATION)


  async function handleButton() {

    if (!session) return router.push(`/auth/login`)

    try {

      setstate('pending')
      
      const res = await addToCart({
        variables: {
          addToCartId: id,
          productId: id
        },
        refetchQueries: [{ query: QUERY_USER_CURRENT }],
      })

      await delay(500)

      setstate('success')

      await delay(500)

      setstate(undefined)
      
    } catch (error) {
      setstate('error')
      console.warn('addtocart error: ', error);
    }


    // console.log({res});

  }

  const renderIcon = (state:State) => {

    switch (state) {
      case 'pending':
        return <TbLoader className={styles.spin}/>
        
      case 'success':
        return <TbCheck />

      case 'error':
        return <TbExclamationCircle />
    
      default:
        return <MdShoppingBag /> 
    }
  }

  return (<>
    <button 
      type="button" 
      disabled={loading || state !== undefined} 
      onClick={e => handleButton()}
      className={' addtocart' + ' button'}
    >
      <span>Add To Cart <span style={{marginRight: 'auto'}} className='state'>{renderIcon(state)}</span> </span>
      
    </button>
    
    </>);
}

const ADD_TO_CART_MUTATION = gql`
  mutation Mutation($addToCartId: ID!, $productId: ID) {
    addToCart(id: $addToCartId, productID: $productId) {
      id
      quantity
      product {
        name
      }
    }
  }
`
