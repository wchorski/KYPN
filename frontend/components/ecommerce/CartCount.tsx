import styled from "styled-components"
import {CSSTransition, TransitionGroup,} from 'react-transition-group';
import { useRef } from "react";
import { MdShoppingBag } from "react-icons/md";

export default function CartCount({count}:{count:number}) {

  const nodeRef = useRef(null);

  return (
    <StyledAnimation>
      <TransitionGroup>
        <CSSTransition
          nodeRef={nodeRef}
          className={'count'}
          classNames={'count'}
          key={count}
          timeout={{enter: 400, exit: 400}}
        >
          <StyledDot ref={nodeRef}>
            {count}
            <MdShoppingBag />
          </StyledDot>
        </CSSTransition>
      </TransitionGroup>
    </StyledAnimation>
  )
}

const StyledDot = styled.div`
  background-color: var(--c-1);
  color: var(--c-txt-rev);
  padding: 0.5rem;
  border-radius: 50px;
  line-height: 2rem;
  min-width: 3rem;
  margin-left: 1rem;
  font-feature-settings: 'tnum';
  font-variant-numeric: tabular-nums;

`
// TODO this animation is wack
const StyledAnimation = styled.span`
  position: relative;

  .count{
    display: block;
    position: relative;
    transition:  0.4s;
    backface-visibility: hidden;
    transition: 0.4s;
    /* position: absolute; */
    top: 0;
    left: 0;
  }

  .count-enter{
    /* background-color: green; */
    opacity: 0;
    transform: scale(1.3) rotateX(45deg);
  }

  .count-enter-active {
    /* background-color: red; */
    opacity: 1;
    /* transition: 2s; */
    /* transform: rotateX(0); */
  }

  .count-exit{
    /* background-color: blue; */
    opacity: 0;
    top: 0;
    position: absolute;
    /* transform: rotateX(0); */
  }

  .count-exit-active{
    /* background-color: pink; */
    opacity: 1;
    /* transition: 2s; */
    transform: scale(1) rotateX(0);
  }
`