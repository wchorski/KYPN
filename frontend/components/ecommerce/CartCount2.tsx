import styled from "styled-components"
import { MdShoppingBag } from "react-icons/md";


export function CartCount2({ count }: { count: number }) {

  return (
    <StyledCartIcon className="cartCount">

      <StyledDot key={count}>
        <span> {count} </span>
      </StyledDot>

      <MdShoppingBag />

    </StyledCartIcon>
  )
}


const StyledDot = styled.div`

  background-color: var(--c-txt-rev);
  color: var(--c-txt);
  padding: 0.2rem;
  position: absolute;
  top: -10px;
  right: 0;
  border-radius: 100px;
  line-height: 2rem;
  min-width: 2.5rem;
  font-feature-settings: 'tnum';
  font-variant-numeric: tabular-nums;
  animation: bigSmall .3s;
  animation-timing-function: cubic-bezier(0.4, 0, 0.2, 1);
  outline: solid white 1px;


  @keyframes bigSmall {
    0% { 
      font-size: 25px;
      padding: 1rem;
    }
    100% { 
      font-size: 2px;
      padding: 0.2rem;
    }
  }

  span {

    transition: all .3s;
  }
`

const StyledCartIcon = styled.div`
  position: relative;
`