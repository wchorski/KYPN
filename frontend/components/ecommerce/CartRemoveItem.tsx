import { gql, useMutation } from "@apollo/client"
import styled from "styled-components"


export default function CartRemoveItem({id}:{id:string}) {

  const [mutate, {loading, error, data}] = useMutation(MUTATE_CART_REMOVE_ITEM)

  async function handleMutation() {
    const res = await mutate({
      variables: {
        where: {
          id: id
        }
      },
      update: handleUpdate,
      // TODO ep 50, will come back to this. fix cache issues
      // optimisticResponse: {
      //   __typename: 'CartItem',
      //   id: id,
      // }
    })

    // console.log({res})
    
  }

  function handleUpdate(cache:any, payload:any) {
    cache.evict(cache.identify(payload.data.deleteCartItem))
  }
    
  return (
    <StyledBigButton 
      type="button" 
      title='Remove this item from cart'
      disabled={loading}
      onClick={handleMutation}
    >
      &times;
    </StyledBigButton>
  )
}

const MUTATE_CART_REMOVE_ITEM = gql`
  mutation DeleteCartItem($where: CartItemWhereUniqueInput!) {
    deleteCartItem(where: $where) {
      id
      quantity
    }
  }
`

const StyledBigButton = styled.button`
  font-size: 3rem;
  background: none;
  border: 0;
  color: var(--c-desaturated);

  &:hover, &:focus{
    color: var(--c-error);
  }
`