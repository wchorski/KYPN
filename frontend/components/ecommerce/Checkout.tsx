import { StyledSickButton } from "@/styles/SickButton.styled"
import { CardElement, Elements } from "@stripe/react-stripe-js"
import { loadStripe } from "@stripe/stripe-js"
import { FormEvent } from "react"
import styled from "styled-components"

// TODO Add blocker stops stripe.com requests thinks it's X site
const stripeLib = loadStripe(process.env.NEXT_PUBLIC_STRIPE_KEY || 'NO KEY IN ENV')

export default function Checkout() {

  function handleSubmit(e:FormEvent){
    e.preventDefault()
    console.log('check me out');
    
  }

  return (
    <Elements stripe={stripeLib}>
      <StyledCheckoutForm onSubmit={e => handleSubmit(e)}>
        
        <CardElement />

        <StyledSickButton>
          Checkout
        </StyledSickButton>
        
      </StyledCheckoutForm>
    </Elements>
  )
}

const StyledCheckoutForm = styled.form`
  box-shadow: 0 1px 2px 2px black;
  background-color: azure;
  border: 1px solid black;
  border-radius: 5px;
  padding: 1rem;
  display: grid;
  grid-gap: 1rem;

`