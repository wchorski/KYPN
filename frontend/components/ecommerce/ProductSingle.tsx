import { QueryError } from '@/components/menus/QueryError';
import { QueryLoading } from '@/components/menus/QueryLoading';
import { ProductThumbnail } from '@/components/ProductThumbnail';
import moneyFormatter from '@/lib/moneyFormatter';
import { gql, useQuery } from '@apollo/client';
import Image from 'next/image';
import ErrorMessage from '../ErrorMessage';
import { StyledPriceTag } from "@/styles/PriceTag.styled";
import Head from 'next/head';
import styled from 'styled-components';


export default function ProductSingle({id}: any) {

  const { loading, error, data } = useQuery(
    SINGLE_PRODUCT_QUERY, {
    variables: { where: { id: id}}
  })
  console.log(data);

  const {name, photo, price, status, description} = data.product
  
  
  if (loading) return <QueryLoading />
  if (error) return <ErrorMessage error={error} />

  return (<>

    <Head>
      <title>Sick Fits | {name}</title>
    </Head>

    <StyledProductPage>
      <Image src={photo.image.url} alt={photo.altText} width={photo.image.width} height={photo.image.height}/>
      
      <div className="details">
        <h2>{name}</h2>

        <span><StyledPriceTag> {moneyFormatter(price)} </StyledPriceTag></span>
        <span>status: {status}</span>

        <p>{description}</p>
      </div>

    </StyledProductPage>
  </>)
}

const SINGLE_PRODUCT_QUERY = gql`
  query Query($where: ProductWhereUniqueInput!) {
    product(where: $where) {
      id
      name
      photo {
        altText
        id
        image {
          extension
          filesize
          id
          height
          url
          width
        }
      }
      price
      status
      description
    }
  }
`

const StyledProductPage = styled.article`
  position: relative;
  display: grid;
  grid-auto-columns: 1fr;
  grid-auto-flow: column;
  min-height: 800px;
  max-width: var(--maxWidth);
  align-items: top;
  gap: 2em

  img{
    width: 100%;
    object-fit: contain;
  }
`