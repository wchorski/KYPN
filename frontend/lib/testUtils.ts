import casual from 'casual';
import { QUERY_PRODUCTS_COUNT } from '../components/Pagination';

// seed it so we get consistent results
casual.seed(777);

const fakeItem = () => ({
  // __typename: 'Item',
  id: 'abc123',
  price: 5000,
  user: null,
  photo: {
    id: 'abc123',
    altText: 'dogs are best',
    image: {
      url: 'dog.jpg',
    },
  },
  name: 'dogs are best',
  description: 'dogs',
});

const fakeUser = (overrides: any | void) => ({
  __typename: 'User',
  id: '4234',
  name: casual.name,
  email: casual.email,
  permissions: ['ADMIN'],
  orders: [],
  cart: [],
  ...overrides,
});

const fakeOrderItem = () => ({
  __typename: 'OrderItem',
  id: casual.uuid,
  image: {
    image: `${casual.word}.jpg`,
  },
  name: casual.words(),
  price: 4234,
  quantity: 1,
  description: casual.words(),
});

const fakeOrder = () => ({
  __typename: 'Order',
  id: 'ord123',
  charge: 'ch_123',
  total: 40000,
  items: [fakeOrderItem(), fakeOrderItem()],
  createdAt: '2022-12-11T20:16:13.797Z',
  // @ts-ignore
  user: fakeUser(),
});

const fakeCartItem = (overrides: any | void) => ({
  __typename: 'CartItem',
  id: 'omg123',
  quantity: 3,
  product: fakeItem(),
  // @ts-ignore
  user: fakeUser(),
  ...overrides,
});

// Fake LocalStorage
class LocalStorageMock {
  constructor() {
    // @ts-ignore
    this.store = {};
  }

  clear() {
    // @ts-ignore
    this.store = {};
  }

  getItem(key: any) {
    // @ts-ignore
    return this.store[key] || null;
  }

  setItem(key: any, value: any) {
    // @ts-ignore
    this.store[key] = value.toString();
  }

  removeItem(key: any) {
    // @ts-ignore
    delete this.store[key];
  }
}

function makePaginationMocksFor(length: any) {
  return [
    {
      request: { query: QUERY_PRODUCTS_COUNT },
      result: {
        data: {
          _allProductsMeta: {
            count: length,
          },
          itemsConnection: {
            __typename: 'aggregate',
            aggregate: {
              count: length,
              __typename: 'count',
            },
          },
        },
      },
    },
  ];
}

export {
  makePaginationMocksFor,
  LocalStorageMock,
  fakeItem,
  fakeUser,
  fakeCartItem,
  fakeOrder,
  fakeOrderItem,
};
