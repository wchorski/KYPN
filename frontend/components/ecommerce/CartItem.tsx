import moneyFormatter from "../../lib/moneyFormatter"
import { StyledCartItem } from "../../styles/CartItem.styled"
import Image from "next/image"
import CartRemoveItem from "./CartRemoveItem"
import { handlePhoto } from "../../lib/handleProductPhoto"
import { ImageDynamic } from "../elements/ImageDynamic"

export default function CartItem({ item }: any) {


  if (!item?.product) return (
    <StyledCartItem>
      <p>This cart item is no longer supplied by our store</p>
    </StyledCartItem>
  )


  const { product: { id, description, name, price, photo, }, quantity } = item


  return (
    <StyledCartItem>
      {/* <Image
        priority
        src={handlePhoto(photo).image?.url}
        alt={handlePhoto(item.photo).image?.altText ? handlePhoto(item.photo).image?.altText : 'no product photo'}
        width={handlePhoto(photo).image?.width}
        height={handlePhoto(photo).image?.height}
      /> */}

      <ImageDynamic photoIn={photo} />

      <h5>{name}</h5>

      <div className="perItemTotal">
        <p>{moneyFormatter(price * quantity)}</p>

        <em>{quantity} &times; {moneyFormatter(price)} each </em>
      </div>

      <CartRemoveItem id={item.id} />
    </StyledCartItem>
  )
}