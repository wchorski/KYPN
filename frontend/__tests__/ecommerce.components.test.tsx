import { render, screen, waitFor } from '@testing-library/react'
import { fakeItem } from '../lib/testUtils'
import { ProductSingle, SINGLE_PRODUCT_QUERY } from '../components/ecommerce/ProductSingle'
import { MockedProvider } from '@apollo/react-testing'
import { withRouter } from "next-router-mock";
import CartCount from '../components/ecommerce/CartCount';

jest.mock('next/dist/client/router', () => require('next-router-mock'));



const fakeProduct = fakeItem()

const myMocks = [
  {
    request: {
      query: SINGLE_PRODUCT_QUERY,
      variables: {
        id: '123',
      }
    },

    result: {
      data: {
        Product: fakeProduct
      }
    }

  }
]

describe('<ProductSingle />', () => {
  // it('render proper query data, ', async () => {

  //   const ProductSingleWrap = withRouter(ProductSingle)

  //   const { container, debug } = render(
  //     <MockedProvider mocks={myMocks} addTypename={false}>
  //       <ProductSingleWrap product={fakeProduct} />
  //     </MockedProvider>
  //   )

  //   await waitFor(() => {
  //     // debug()
  //     screen.findByTestId('singleProduct')
  //     expect(container).toMatchSnapshot()
  //   })
  // })

  it('apollo error if item not found, ', async () => {
    const errorMocks: any = [{
      request: {
        query: SINGLE_PRODUCT_QUERY,
        variables: {
          where: {
            id: '123',
          }
        }
      },
      result: {
        errors: [{ message: 'Items not found' }]
      }
    }]

    const ProductSingleWrap = withRouter(ProductSingle)

    const { container, debug } = render(
      <MockedProvider mocks={errorMocks} >
        <ProductSingleWrap id='123' />
      </MockedProvider>
    )

    await waitFor(() => {
      screen.findByTestId('graphql-error')
      expect(container).toHaveTextContent('Items not found')
      // debug()
    })
  })
})

// describe('<ProductSingle />', () => {

//   const mocks: any = []

//   it('renders price tag & title', async () => {

//     const ProductSingleWrap = withRouter(ProductSingle)

//     const { container, debug } = render(
//       <MockedProvider mocks={mocks} addTypename={false}>
//         <ProductSingleWrap product={fakeProduct} />
//       </MockedProvider>
//     )


//     await waitFor(() => {
//       // debug()
//       const img = screen.getByAltText(fakeProduct.name);
//       expect(img).toBeInTheDocument();

//       const priceTag = screen.getByText('$50');
//       expect(priceTag).toBeInTheDocument();

//       const link = container.querySelector('a');
//       expect(link).toHaveAttribute('href', '/shop/product/update?id=abc123');
//       expect(link).toHaveTextContent('Edit ');

//     })
//   })

//   it('renders photo properly', async () => {
//     const ProductSingleWrap = withRouter(ProductSingle)

//     const { container, debug } = render(
//       <MockedProvider mocks={mocks} addTypename={false}>
//         <ProductSingleWrap product={fakeProduct} />
//       </MockedProvider>
//     )


//     await waitFor(() => {

//     })
//   })
// })


describe('<CartCount />', () => {
  // useRouter.mockReturnValue({ query: {} })
  const mocks: any = []

  it('renders cart count', async () => {


    const { container, debug } = render(
      <MockedProvider mocks={mocks} addTypename={false}>
        <CartCount count={23} />
      </MockedProvider>
    )

    await waitFor(() => {

      // debug()
    })
  })

  it('matches snapshot', () => {

    const { container } = render(
      <CartCount count={11} />
    )

    expect(container).toMatchSnapshot()
  })

  it('updates via props', async () => {

    const { container, rerender, debug } = render(
      <CartCount count={11} />
    )

    await waitFor(() => {
      expect(container.textContent).toBe('11')
      rerender(<CartCount count={12} />)
      expect(container).toMatchSnapshot()
      // debug()
    })

  })
})



