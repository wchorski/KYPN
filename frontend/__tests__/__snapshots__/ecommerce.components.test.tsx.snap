// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<CartCount /> matches snapshot 1`] = `
<div>
  <span
    class="sc-f8e340b4-1 BPwIA"
  >
    <div>
      <div
        class="sc-f8e340b4-0 gBYRFW count"
      >
        11
        <svg
          fill="currentColor"
          height="1em"
          stroke="currentColor"
          stroke-width="0"
          viewBox="0 0 24 24"
          width="1em"
          xmlns="http://www.w3.org/2000/svg"
        >
          <path
            d="M0 0h24v24H0z"
            fill="none"
          />
          <path
            d="M18 6h-2c0-2.21-1.79-4-4-4S8 3.79 8 6H6c-1.1 0-2 .9-2 2v12c0 1.1.9 2 2 2h12c1.1 0 2-.9 2-2V8c0-1.1-.9-2-2-2zm-8 4c0 .55-.45 1-1 1s-1-.45-1-1V8h2v2zm2-6c1.1 0 2 .9 2 2h-4c0-1.1.9-2 2-2zm4 6c0 .55-.45 1-1 1s-1-.45-1-1V8h2v2z"
          />
        </svg>
      </div>
    </div>
  </span>
</div>
`;

exports[`<CartCount /> updates via props 1`] = `
<div>
  <span
    class="sc-f8e340b4-1 BPwIA"
  >
    <div>
      <div
        class="sc-f8e340b4-0 gBYRFW count count-enter count-exit count-enter-active count-exit-active"
      >
        12
        <svg
          fill="currentColor"
          height="1em"
          stroke="currentColor"
          stroke-width="0"
          viewBox="0 0 24 24"
          width="1em"
          xmlns="http://www.w3.org/2000/svg"
        >
          <path
            d="M0 0h24v24H0z"
            fill="none"
          />
          <path
            d="M18 6h-2c0-2.21-1.79-4-4-4S8 3.79 8 6H6c-1.1 0-2 .9-2 2v12c0 1.1.9 2 2 2h12c1.1 0 2-.9 2-2V8c0-1.1-.9-2-2-2zm-8 4c0 .55-.45 1-1 1s-1-.45-1-1V8h2v2zm2-6c1.1 0 2 .9 2 2h-4c0-1.1.9-2 2-2zm4 6c0 .55-.45 1-1 1s-1-.45-1-1V8h2v2z"
          />
        </svg>
      </div>
      <div
        class="sc-f8e340b4-0 gBYRFW count"
      >
        11
        <svg
          fill="currentColor"
          height="1em"
          stroke="currentColor"
          stroke-width="0"
          viewBox="0 0 24 24"
          width="1em"
          xmlns="http://www.w3.org/2000/svg"
        >
          <path
            d="M0 0h24v24H0z"
            fill="none"
          />
          <path
            d="M18 6h-2c0-2.21-1.79-4-4-4S8 3.79 8 6H6c-1.1 0-2 .9-2 2v12c0 1.1.9 2 2 2h12c1.1 0 2-.9 2-2V8c0-1.1-.9-2-2-2zm-8 4c0 .55-.45 1-1 1s-1-.45-1-1V8h2v2zm2-6c1.1 0 2 .9 2 2h-4c0-1.1.9-2 2-2zm4 6c0 .55-.45 1-1 1s-1-.45-1-1V8h2v2z"
          />
        </svg>
      </div>
    </div>
  </span>
</div>
`;
